### PR TITLE
Add the CLZ and POPCNT esil operations ##esil

### PR DESCRIPTION
**Pull request by phix33**
- [X] Mark this if you consider it ready to merge
- [X] I've added tests (optional)
- [ ] I wrote some lines in the [book](https://github.com/radareorg/radare2book) (optional)

**Description**

Adds two esil operations and pays them off in three arches at once:

- CLZ — w,x,CLZ pushes the count of leading zeros of x within its low w bits (w when x is 0). The value is masked to the width inside the op, so callers need no pre-masking, and the width is an explicit operand following ~ — not taken from the operand or the config, which is what keeps the register-width ops from expressing sub-width operations.
- POPCNT — x,POPCNT pushes the number of set bits.

Both are backed by the portable helpers in r_util/r_bits.h, registered in the esil DFG, documented in ae??, and pinned by direct ae tests including the edge cases (zero value, width clamp, high-garbage masking).

 Consumers, one commit each:

- ppc — cntlzw/cntlzd stop being type-only, and popcntb/popcntw/popcntd are modelled (per-byte and per-word lanes built from POPCNT).  The record form cntlzw. picks up cr0 through the existing generic Rc handling. Verified against the qemu differential. 
- arm32 — the GOTO-loop clz is deleted: it clobbered its destination as scratch and derived its jump target by counting commas in the emitted buffer. Now 32,rm,CLZ,rd,=, still under the conditional prefix.
- arm64 — the 40-line unrolled binary-search clz is deleted for w,rn,CLZ,rd,= at the register width; it also stops using tmp as scratch.

All arm suites of the unicorn differential are unchanged too.

This unblocks the same modelling for mips clz/clo/dclz/dclo (follows separately — the mips esil switch is being touched by an open PR) and would let x86's popcnt/lzcnt drop their SWAR and smear loops later. Also, the two remaining bug groups the extended qemu differential found:

* arm64 extended-register operands — the last wrong-value group: uxtb/sxth extenders dropped, bfxil/extr/smulh wrong, smsubl NOESIL, plus the core ~/ INT_MIN/-1 edge.
* rasm2 assembler bugs — LOW: mis-encodings the harness routes around via llvm-mc, so they only bite rasm2/wa users.

### CODE REVIEW

**Comment by trufae** [comment]
Personally im all for adding new esil ops. Despite we must review and remove a bunch of them already those two can be also reimplemented in plain esil with the cost of extra esilvm cpu time which is bad for emulation. But i think those cases maybe are quite commonly found across archs to really matter having yhem natively in esil..

**Comment by trufae** [comment]
I reviewed the patch and, taking @trufae's comment as the starting point, tried to answer the underlying question: when does an operation deserve a native ESIL op instead of plain ESIL. Verdict up front: the direction is right and the patch is merge-worthy after a rebase.

## Review of the patch itself

- `esil_clz`/`esil_popcount` follow the current `RStrs` idiom, and the CLZ math is correct: the value is masked to the width first so `clz64(val) - (64 - width)` cannot underflow, `r_num_bitmask(64)` covers the full-width case, width is clamped to 64, and width 0 degenerates cleanly. The `width,value,CLZ` operand order matches the existing `count,value,ASR` convention, and the DFG registration is included — `esil_dfg.c` is the only other in-tree table that enumerates ops, so nothing else needs teaching.
- I hand-evaluated the ppc `popcntw`/`popcntb` lane compositions on the stack: both are correct, including `rd == rs`, since everything is computed before the assignment.
- The deleted implementations deserved to go: the arm32 clz derived its GOTO target by counting commas in the emitted buffer, and both arm versions used scratch registers — the same clobber class that already produced the `rev x0,x0` bug the REV block comments on.

Non-blocking findings:

- The conflict with master is only append-collisions at the end of `test/db/esil/arm_64` and `test/db/esil/esil`, from the arm64 esil fixes that landed after this branch. A rebase resolves it mechanically.
- `PPC_INS_POPCNTB` is guarded with `CS_API_MAJOR > 4` but `POPCNTW`/`POPCNTD` are not, and CI has no capstone-v4 job — worth confirming v4's `ppc.h` has those enums before merge.
- `plugin_v35.c:1750` still carries its own copy of the unrolled arm64 clz; cheap follow-up.
- POPCNT taking no width operand is the right asymmetry (popcount of a masked value is width-independent, CLZ's result is not), but the rule "width operand only where semantics depend on it" should be written down in the book so future ops stay consistent.

## When a native op is justified

Surveying the tree, the pure-ESIL versions of this op family fall into two classes: GOTO loops (arm32 clz deleted here, arm64 `rbit`, x86 `bsf`/`bsr` — all with comma-counted jump heads) and large unrolled expressions (x86 SWAR popcount and smear-loop lzcnt, the byte-swap chains duplicated across x86/arm64/arm32). The loops are the killer: beyond the 30–60 VM steps per instruction, they defeat `esil_dfg` entirely and their jump-head arithmetic breaks whenever a conditional prefix changes the token count. So the bar I would set, which CLZ and POPCNT both pass:

1. plain ESIL needs a loop or a >20-token unrolled expression;
2. it recurs across 3+ architectures;
3. it maps to one portable helper (one C statement, here `r_bits.h`);
4. fixed stack arity so the DFG and static consumers stay trivial.

The counterweight is that every op is permanent language surface: external ESIL consumers must learn it, and once an arch plugin emits it we can never remove it. So the bar should stay "loop-or-buggy, and multi-arch" — not "shorter than five tokens".

## ESIL macros as the middle path

For the cases that are expressible in plain ESIL but boilerplate-heavy, it may be worth reintroducing the concept of ESIL macros — the Forth-style `: word ... ;` definitions already prototyped in `libr/esil/esil_compiler.c`. A centrally defined macro that expands to plain ESIL would let arch plugins emit one readable token without growing the interpreter core, give every op a canonical reference expansion (which pins edge-case semantics and keeps external consumers implementable), and reserve native C ops for the loop-or-inexpressible cases. Native op for speed where it matters, macro for expressiveness where it doesn't — that keeps the op count from creeping.

## The bit ops we will need next if we take this direction

- **CTZ** (`w,x,CTZ`) — the symmetric twin; `r_bits_ctz64` already sits unused in `r_bits.h`. Turns x86 `bsf` from a GOTO loop into a zero-guard plus one op, makes `bsr` a `w-1-CLZ`, and models riscv B-ext `ctz`.
- **BSWAP** (`w,x,BSWAP`) — the most duplicated idiom in the tree: x86 `bswap`/`movbe` (`byteswap_expr`), arm64 REV/REV16/REV32, arm32 rev/rev16/revsh, ppc byte-reversed load/stores. Also removes the `tmp` scratch usage in the arm64 REV block.
- **RBIT** (`w,x,RBIT`) — arm64 `rbit` is the last comma-counted GOTO loop in the arm plugin; riscv `brev8` too. Genuinely inexpressible without a loop, so a macro cannot cover it.

After those three the well runs mostly dry. Wide multiply already exists (`L*`). Saturating arithmetic (arm DSP QADD family) recurs but has too many signed/width variants — better to wait until differential testing reaches it before fixing semantics. Parity (`$p`), sign-extend (`~`), abs/min/max, and single-arch cases like x86 `pdep`/`pext` don't clear the bar and are macro or plain-ESIL material.

On the "we must review and remove a bunch of them" point: the bloat is in the ~60 mechanical composite memory ops (`+=[N]`, `<<=[N]`, …) and duplicate aliases (LSL/LSR vs `<<`/`>>`), not the math primitives — adding these few mnemonic bit ops while pruning that family shrinks the total surface.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by phix33** [comment]
Rebased, and documented both ops in doc/esil.md and man/esil.7.

The capstone guard is deliberate: v4 has POPCNTD, POPCNTW, CNTLZW and CNTLZD, and lacks only POPCNTB. Guarding the other four would  make them unreachable on v4.

#26556 is a better answer to the op count than anything here (I had started planing but go side-tracked with other things today - sorry!). In any case, they're orthogonal — CLZ/POPCNT cover the cases plain ESIL needs a loop for, like the arm32 clz this deletes, which found its jump target by counting commas in the emitted buffer. Happy to rebase onto your stack.

**Comment by trufae** [comment]
please rebase, i have implemented support for esil defines, to implement words/operations as plain esil strings  instead of C. this way we keep the design simple and we are a bit closer to forth now

File: libr/esil/esil_ops.c [hunk 1]
```diff
@@ -113,6 +114,34 @@ static ut32 esil_internal_packed_size_reg(REsil *esil, const char *r) {
...
+	ut64 val, width;
+	const RStrs p_val = r_esil_pop (esil);
+	if (r_strs_empty (p_val) || !r_esil_get_parm (esil, p_val, &val)) {
+		return false;
+	}
+	const RStrs p_width = r_esil_pop (esil);
+	if (r_strs_empty (p_width) || !r_esil_get_parm (esil, p_width, &width)) {
+		return false;
+	}
+	if (width > 64) {
```
Comment (trufae) [inline libr/esil/esil_ops.c:128]: maybe R_LOG_WARN or TRAP because thats not a valid thing. funnily this falls in the line128 lel

File: libr/arch/p/arm/plugin_cs.c [hunk 2]
```diff
@@ -1942,48 +1942,9 @@ static int analop64_esil(RArchSession *as, RAnalOp *op, ut64 addr, const ut8 *bu
...
-			"tmp,0xffff0000,&,!,?{,16,tmp,<<=,16,+,},"
-			"tmp,0xff000000,&,!,?{,8,tmp,<<=,8,+,},"
-			"tmp,0xf0000000,&,!,?{,4,tmp,<<=,4,+,},"
-			"tmp,0xc0000000,&,!,?{,2,tmp,<<=,2,+,},"
-			"tmp,0x80000000,&,!,?{,1,+,},"
-			"%s,!,?{,32,%s,=,}{,%s,=,}",
-			r1, r1, r0, r0);
-		} else {
-			r_strbuf_setf (&op->esil,
-			"%s,tmp,=,0,"
```
Comment (trufae) [inline libr/arch/p/arm/plugin_cs.c:1975]: despite the old implementation uses tmp fake register and is very verbose. i think it will be good to keep it under `#if 0` block for r_esil_define, mainly because we know that radius, radeco, r2smt and meny other software using esil expressions wont work if they use the new instructions you added.. thats why is tricky to add/replace/remove esil ops because we have no real deprecation plan and i understand those two primitives are good enough to implement the rest of bitops without too much mess but now that we have defines support maybe we can do some fallbacks and start caring about future plans in esil

**Review by trufae — CHANGES_REQUESTED** [review]
im fine to merge because ive been thinking about this problem for years and im tired of not deciding a good path towards having those ops implemented in a portable, efficient and clean way across multiple archs, and having these ops is probably the way to go. i was waiting for @condret feedback but we can always merge and fallback with a define in any case, because the new semantics are cleaner

**Comment by trufae** [comment]
I went through the review comments and re-checked the code on cb938e1 against them. Verdict up front: the implementations are correct as they stand, CI is green on the current head, and both inline comments are right to raise — one resolves with a two-line diagnostic, and for the fallback/deprecation concern I'd propose a different home than `#if 0` in the plugin. Details below.

## `esil_ops.c:128` — silent clamp of `width > 64` (WARN or trap)

Agreed it should not be silent. Two notes on the specifics:

- The clamp has to stay regardless of the diagnostic: `r_num_bitmask()` takes a `ut8`, so an unclamped width of e.g. 256 would wrap to mask 0, and `clz64 (val) - (64 - width)` underflows for nonzero values. The only question is what to put on top of it.
- The file already has both precedents: `esil_lsl` does `R_LOG_DEBUG ("shift is too big")` and fails the op; `esil_div` sets `esil->trap = R_ANAL_TRAP_DIVBYZERO`, pushes 0 and continues.

I'd go with `R_LOG_WARN` + clamp-and-continue rather than a trap: an oversized width can only come from an arch-plugin bug or a hand-typed `ae` (every in-tree caller passes a literal constant), so it's an authoring-time error worth being loud about — while traps are the tool for data-dependent conditions the guest can legitimately reach, and trapping here would turn a decode bug into a dead emulation session. If a trap is preferred anyway, the `ae 70,1,CLZ → 0x3f` expectation in `test/db/esil/esil` is the one test that changes. (And indeed 128 is the first width that realistically triggers it, on its own line number.)

## `plugin_cs.c` — keeping the old expansions under `#if 0` for `r_esil_define`

Agree with the goal (a plain-esil fallback and a deprecation story), but I'd argue against preserving these two bodies, and against the plugin as the location:

- These specific expansions are the wrong reference to keep. The arm32 one derives its GOTO target by counting commas in the emitted buffer (breaks the moment the conditional prefix changes the token count) and uses rd as scratch mid-computation; the arm64 one clobbers `tmp` — the same clobber class as the `rev x0,x0` bug noted next to the REV block. Git history keeps them if we ever need to look.
- More fundamentally, the arm32 body can never become a define. I checked the interpreter: `?{` / `}{` / `}` work fine inside a define body (skip handling lives in `runword` itself, so conditionals nest), but GOTO does not — `esil->parse_goto` is resolved by `eval_word` against the *outer* expression string (`goto_word (ostr, ...)`), so a GOTO issued inside a define body jumps within the caller after the body returns. Loops cannot live in define bodies; only straight-line and conditional code can.
- What *can* be a define is a width-generic, stack-only CLZ: mask, shift left by `64-width`, then the six-step `?{}` binary search juggled on the stack — no scratch register, because a shared define writing `tmp` would reintroduce the clobber bug for every arch at once. Same for a SWAR POPCNT body. If we want the fallback compiled in, the natural home is `esil_ops.c` next to the registration:

  ```c
  #if 1
  	ret &= OP ("CLZ", esil_clz, 1, 2, OT_MATH);
  #else
  	ret &= OPD ("CLZ", "<stack-only body>", 1, 2, OT_MATH);
  #endif
  ```

  One toggle then covers every arch that emits the op, and the ops table already supports replacement in both directions (`r_esil_set_op` explicitly drops a previous define's expansion when overriding it). I'm happy to write and test that body as a follow-up.

## The external-consumer question (radius, radeco, r2smt)

Those projects run their own esil engines over the `op->esil` text, so neither an `#if 0` block nor an in-r2 define reaches them — what reaches them is the language definition. Two concrete things help:

- `ae???j` already dumps the live op table (name, inputs, outputs, info), and define-backed ops carry their full expansion in `info`. That's a machine-readable sync point external engines can pin against. CLZ/POPCNT are currently registered with a NULL info; registering them via `OP2` with a one-liner (`"w,x,CLZ: leading zeros of x in its low w bits, w if x is 0"`) would make `ae???` self-documenting. Only `[1]` and `NAN` use `OP2` today, but this seems like the direction worth standardizing.
- The defines suggest a natural policy going forward: a new compound or derivable op enters as an `OPD` whose body *is* its spec (visible in `ae???j`), and native C is the override for ops that are hot or inexpressible in straight-line esil (CLZ over an arbitrary width; RBIT when it comes). That, plus a release-notes line per added op, is about as much deprecation plan as a single-implementation language can carry.

Both docs added in this PR (doc/esil.md, man/esil.7) describe the two ops and the "width operand only when the result depends on it" rule, so the consumer-facing story is at least written down now.

## Everything else re-checked on cb938e1

- The ppc lane compositions hand-evaluate correctly under the stack semantics (`dst = pop; src = pop`): popcntw's per-word and popcntb's per-byte forms are right, including `rd == rs`, since both source reads happen before the assignment.
- arm64 `8 * REGSIZE64 (0)` yields 32 for wN and 64 for xN, so the width operand is correct in both forms, and the arm32 form still rides under its conditional prefix (the renamed test pins exactly that).
- DFG arities match the op registrations (CLZ consume-2-push-1, POPCNT consume-1-push-1), and the capstone guard is correct as argued before: v4 has all the ppc enums except POPCNTB, which is the only one guarded.

So: +1 to merge once the WARN lands on the clamp; the `OP2` info strings and the esil_ops.c-side OPD fallback can ride along or come as a follow-up, whichever is preferred.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by phix33** [comment]
Rebased onto master and updated:

- `esil_ops.c` — the `width > 64` clamp now logs `R_LOG_WARN ("CLZ width %d clamped to 64")` and continues with every in-tree emitter passing a literal 32/64, so this can only come from a plugin bug or a hand-typed `ae`, and `r_num_bitmask` takes a `ut8` so the clamp itself has to stay. We could switch to a trap if you prefer; `ae 70,1,CLZ -> 0x3f` in `test/db/esil/esil` is the one golden that would move.
- Fallbacks for `r_esil_define` — kept, but in `esil_ops.c` next to the registration rather than the old arm bodies under `#if 0` in the plugin. Two reasons: the deleted arm32 body found its GOTO target by counting commas in the emitted buffer and GOTO cannot live in a define body (it is resolved against the outer expression), and the arm64 one clobbered `tmp`. Instead there is an `#if 1 // ESIL_NATIVE_BITOPS` toggle whose `#else` side registers both ops as plain-esil `OPD` bodies — stack-only (no scratch register, so no clobber for any arch), straight-line binary search for `CLZ` (99 tokens) and the SWAR form for `POPCNT`. I built both sides: the define bodies give the same results as the native ops on 24 edge cases (widths 0/1/17/33/63/64, zero inputs, all-ones) and pass `db/esil/{esil,arm_32,arm_64,ppc_32,ppc_64}` + `db/anal/ppc` unchanged, except one cosmetic golden where a define pushes the literal `32` and the native op prints `0x20`.
- Both ops are now registered with `OP2` info strings, so `ae???` documents them (`w,x,CLZ: leading zeros of x in its low w bits, w if x is 0`, `x,POPCNT: number of set bits in x`) — external engines have the contract in the op table as well as in `doc/esil.md` / `man/esil.7`.

On the wider ESIL issues you raised (no deprecation plan, external consumers, the op table growing), Fable suggested the following — the `#if` block above being the special case of the first point:

- **Reference bodies on native ops.** Let a native op carry a plain-esil body too (`ae???j` shows `body` + `native`), with `e esil.native=false` swapping every such op to its define. C≡body can then be checked over the whole differential corpus with one config flip instead of by hand.
- **Emit-side lowering.** `e esil.lower=true` expands every op that has a body into the primitive set in the `esil` field of `aoj`/`pdj`/`pij`. radius2 reads ESIL from `pdj`, so it never sees a new op — no change on its side.
- **Write it down.** The primitive set and the add/deprecate rules go into `esil.7`, so an op PR is checked against a rule instead of re-opening the debate.
- **Two things the defines changed silently, fixed first:** #26567 — `esil.gotolimit` counted every word including define-body tokens (40 define-backed `CLZ` trap at the default 4096); #26568 — the per-expression arena is sized for native pushes, and a define pushes every intermediate (the CLZ body dies at ~700 uses per expression where native goes to ~30000).
- **For scale:** `r_esil_parse` of `32,5,CLZ` is 0.2 µs native vs 7.6 µs as a define, so native stays the implementation and the body is the spec.
- The red CI on this and the other open PRs is master's own (`APK multidex is.j rebasing`, stale `is.j` golden since e6937fb2c2); #26569 fixes it.

Happy to build any of the first three, or open them as an RFC issue with the numbers, if you want that separate from this PR.
